File: src/test/java/no/nav/fo/veilarbregistrering/registrering/bruker/ValideringUtilsTest.java
package no.nav.fo.veilarbregistrering.registrering.bruker;

import no.nav.fo.veilarbregistrering.besvarelse.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static no.nav.fo.veilarbregistrering.registrering.bruker.ValideringUtils.validerBrukerRegistrering;

class ValideringUtilsTest {

    @Test
    void hvisMistetJobbenSaaSkalManHaSvartPaaUtdanning() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                .setDinSituasjon(DinSituasjonSvar.MISTET_JOBBEN)
                .setSisteStilling(SisteStillingSvar.INGEN_SVAR)
                .setUtdanning(UtdanningSvar.INGEN_SVAR)
                .setUtdanningGodkjent(UtdanningGodkjentSvar.NEI)
                .setUtdanningBestatt(UtdanningBestattSvar.NEI)
                .setHelseHinder(HelseHinderSvar.NEI)
                .setAndreForhold(AndreForholdSvar.NEI)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void hvisSisteStillingSpmIkkeErBesvartSkalManViteHvorvidtBrukerErIJobb() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                .setDinSituasjon(DinSituasjonSvar.USIKKER_JOBBSITUASJON)
                .setSisteStilling(SisteStillingSvar.INGEN_SVAR)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalHaSvartPaaUtdanningssporsmalHvisDinSituasjonIkkeErVilFortsetteIJobb() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                .setDinSituasjon(DinSituasjonSvar.MISTET_JOBBEN)
                .setSisteStilling(SisteStillingSvar.INGEN_SVAR)
                .setUtdanning(UtdanningSvar.INGEN_SVAR)
                .setUtdanningBestatt(UtdanningBestattSvar.INGEN_SVAR)
                .setUtdanningGodkjent(UtdanningGodkjentSvar.INGEN_SVAR)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void stillingSkalSamsvaretMedSvaretPaSisteStillingSpm() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering1 = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering()
                .setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                        .setSisteStilling(SisteStillingSvar.HAR_IKKE_HATT_JOBB))
                .setSisteStilling(StillingTestdataBuilder.gyldigStilling());
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering1));

        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering2 = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering()
                .setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                        .setSisteStilling(SisteStillingSvar.HAR_HATT_JOBB))
                .setSisteStilling(StillingTestdataBuilder.ingenYrkesbakgrunn());
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering2));
    }

    @Test
    void sporsmalSkalIkkeVaereNull() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse().setAndreForhold(null));
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void sisteStillingSkalIkkeVaereNull() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setSisteStilling(null);
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalSvarePaaSporsmalOmAndreForhold() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(
                BesvarelseTestdataBuilder.gyldigBesvarelse().setAndreForhold(AndreForholdSvar.INGEN_SVAR)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalSvarePaaSporsmalOmHelse() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(
                BesvarelseTestdataBuilder.gyldigBesvarelse().setHelseHinder(HelseHinderSvar.INGEN_SVAR)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalHaIngenYrkesbakgrunnHvisViVetAtBrukerIkkeHarHattJobb() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistreringUtenJobb().setSisteStilling(StillingTestdataBuilder.gyldigStilling());
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalIkkeSvarePaaSpmOmSisteStillingHvisViVetAtBrukerIkkeHarHattJobb() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistreringUtenJobb().setBesvarelse(
                BesvarelseTestdataBuilder.gyldigBesvarelseUtenJobb().setSisteStilling(SisteStillingSvar.HAR_IKKE_HATT_JOBB)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void skalIkkeSvarePaaSpmOmSisteStillingHvisViAlleredeVetAtBrukerHarHattJobb() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(BesvarelseTestdataBuilder.gyldigBesvarelse()
                .setDinSituasjon(DinSituasjonSvar.MISTET_JOBBEN)
                .setSisteStilling(SisteStillingSvar.INGEN_SVAR)
        );
        validerBrukerRegistrering(ordinaerBrukerRegistrering);

        ordinaerBrukerRegistrering.setBesvarelse(ordinaerBrukerRegistrering.getBesvarelse()
                .setSisteStilling(SisteStillingSvar.HAR_HATT_JOBB)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void valideringSkalGodkjenneGyldigeObjekter() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering();
        validerBrukerRegistrering(ordinaerBrukerRegistrering);
    }

    @Test
    void valideringSkalFeileHvisBesvarelseHarNullfelt() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setBesvarelse(
                BesvarelseTestdataBuilder.gyldigBesvarelse().setAndreForhold(null)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }

    @Test
    void valideringSkalFeileHvisStillingHarNullfelt() {
        OrdinaerBrukerRegistrering ordinaerBrukerRegistrering = OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering().setSisteStilling(
                StillingTestdataBuilder.gyldigStilling().setLabel(null)
        );
        Assertions.assertThrows(RuntimeException.class, () -> validerBrukerRegistrering(ordinaerBrukerRegistrering));
    }
}

File: src/test/kotlin/no/nav/fo/veilarbregistrering/registrering/bruker/ValideringUtilsTest.kt
package no.nav.fo.veilarbregistrering.registrering.bruker

import no.nav.fo.veilarbregistrering.besvarelse.*
import no.nav.fo.veilarbregistrering.besvarelse.BesvarelseTestdataBuilder.gyldigBesvarelse
import no.nav.fo.veilarbregistrering.besvarelse.StillingTestdataBuilder.gyldigStilling
import no.nav.fo.veilarbregistrering.registrering.bruker.OrdinaerBrukerRegistreringTestdataBuilder.gyldigBrukerRegistrering
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class ValideringUtilsTest {
    @Test
    fun `hvis mistet jobben så skal man ha svart på utdanning`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.MISTET_JOBBEN,
                sisteStilling = SisteStillingSvar.INGEN_SVAR,
                utdanning = UtdanningSvar.INGEN_SVAR,
                utdanningGodkjent = UtdanningGodkjentSvar.NEI,
                utdanningBestatt = UtdanningBestattSvar.NEI,
                helseHinder = HelseHinderSvar.NEI,
                andreForhold = AndreForholdSvar.NEI,
            )
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `hvis siste stilling spm ikke er besvart skal man vite hvorvidt bruker er i jobb`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.USIKKER_JOBBSITUASJON,
                sisteStilling = SisteStillingSvar.INGEN_SVAR,
            )
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal ha svart på utdanningsspørsmal hvis din situasjon ikke er vil fortsette i jobb`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.MISTET_JOBBEN,
                sisteStilling = SisteStillingSvar.INGEN_SVAR,
                utdanning = UtdanningSvar.INGEN_SVAR,
                utdanningBestatt = UtdanningBestattSvar.INGEN_SVAR,
                utdanningGodkjent = UtdanningGodkjentSvar.INGEN_SVAR,
            )
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `stilling skal samsvaret med svaret pa siste stilling spm`() {
        val ordinaerBrukerRegistrering1 = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                sisteStilling = SisteStillingSvar.HAR_IKKE_HATT_JOBB,
            ),
            stilling = gyldigStilling()
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering1
            )
        }
        val ordinaerBrukerRegistrering2 = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                sisteStilling = SisteStillingSvar.HAR_HATT_JOBB,
            ),
            stilling = ingenYrkesbakgrunn
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering2
            )
        }
    }

    @Test
    fun `spørsmal skal ikke være null`() {
        val ordinaerBrukerRegistrering =
            gyldigBrukerRegistrering(
                besvarelse = gyldigBesvarelse(andreForhold = null)
            )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal svare på sporsmal om andre forhold`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(andreForhold = AndreForholdSvar.INGEN_SVAR)
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal svare på spørsmal om helse`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(helseHinder = HelseHinderSvar.INGEN_SVAR)
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal ha ingen yrkesbakgrunn hvis vi vet at bruker ikke har hatt jobb`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            stilling = gyldigStilling(),
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.ALDRI_HATT_JOBB,
                sisteStilling = SisteStillingSvar.INGEN_SVAR,
            )
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal ikke svare på spm om siste stilling hvis vi vet at bruker ikke har hatt jobb`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            stilling = ingenYrkesbakgrunn,
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.ALDRI_HATT_JOBB,
                sisteStilling = SisteStillingSvar.HAR_IKKE_HATT_JOBB,
            )
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `skal ikke svare på spm om siste stilling hvis vi allerede vet at bruker har hatt jobb`() {
        var ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(
                dinSituasjon = DinSituasjonSvar.MISTET_JOBBEN,
                sisteStilling = SisteStillingSvar.INGEN_SVAR,
            )
        )
        ValideringUtils.validerBrukerRegistrering(ordinaerBrukerRegistrering)
        ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            opprettetDato = ordinaerBrukerRegistrering.opprettetDato,
            besvarelse = gyldigBesvarelse(
                sisteStilling = SisteStillingSvar.HAR_HATT_JOBB,
                dinSituasjon = DinSituasjonSvar.MISTET_JOBBEN,
            ),
            stilling = ordinaerBrukerRegistrering.sisteStilling,
            teksterForBesvarelse = ordinaerBrukerRegistrering.teksterForBesvarelse,
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `validering skal godkjenne gyldige objekter`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering()
        ValideringUtils.validerBrukerRegistrering(ordinaerBrukerRegistrering)
    }

    @Test
    fun `validering skal feile hvis besvarelse har nullfelt`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            besvarelse = gyldigBesvarelse(andreForhold = null)
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }

    @Test
    fun `validering skal feile hvis stilling har nullfelt`() {
        val ordinaerBrukerRegistrering = gyldigBrukerRegistrering(
            stilling = gyldigStilling(label = null)
        )
        assertThrows<RuntimeException> {
            ValideringUtils.validerBrukerRegistrering(
                ordinaerBrukerRegistrering
            )
        }
    }
}
